src/olympia/amo/middleware.py

```diff
@@ -188,15 +191,50 @@ def process_request(self, request):
 
 
 class ReadOnlyMiddleware(object):
+    """Middleware that announces a downtime which for us usually means
+    putting the site into read only mode.
+
+    Supports issuing `Retry-After` header.
+    """
+    ERROR_MSG = _(
+        u'Some features are temporarily disabled while we '
+        u'perform website maintenance. We\'ll be back to '
+        u'full capacity shortly.')
+
+    API_HEADER_NAME = 'X-AMO-Read-Only'
+
+    def _render_api_error(self):
+        response = JsonResponse({'error': self.ERROR_MSG}, status=503)
+        response[self.API_HEADER_NAME] = 'true'
+        if settings.READ_ONLY_RETRY_AFTER is not None:
+            response['Retry-After'] = settings.READ_ONLY_RETRY_AFTER.seconds
+        return response
 
     def process_request(self, request):
-        if request.method == 'POST':
+        if not settings.READ_ONLY:
+            return
+
+        if request.path.startswith('/api/'):
+            writable_method = request.method in ('POST', 'PUT', 'DELETE')
+            if settings.READ_ONLY and writable_method:
+                return self._render_api_error()
+        elif request.method == 'POST':
             return render(request, 'amo/read-only.html', status=503)
 
     def process_exception(self, request, exception):
```

Comment:
If I understand Middleware classes correctly this is always called for exceptions so we need to test for `settings.READ_ONLY` here too.  (The previous behaviour worked because the entire middleware class was only added when the setting was enabled)

Reply:
Indeed, fixed in 5c566708eb